src/pveclib/vec_int512_ppc.h

```diff
@@ -0,0 +1,2167 @@
+#define __VEC_PWR_IMP(FNAME) FNAME ## _PWR7
+#endif
+#endif
+
+/** \brief Vector Add 512-bit Unsigned Integer & write Carry.
+ *
+ *  Compute the 512 bit sum of two 512 bit values a, b and
+ *  product the carry.
+ *  The sum (with-carry) is returned as single 640-bit integer in a
+ *  homogeneous aggregate structure.
+ *
+ *  |processor|Latency|Throughput|
+ *  |--------:|:-----:|:---------|
+ *  |power8   |  16   | 1/cycle  |
+ *  |power9   |  12   | 1/cycle  |
+ *
+ *  @param a vector representation of a unsigned 512-bit integer.
+ *  @param b vector representation of a unsigned 512-bit integer.
+ *  @return homogeneous aggregate representation of the unsigned
+ *  640-bit sum of a + b.
+ */
+static inline __VEC_U_640
+vec_add512cu (__VEC_U_512 a, __VEC_U_512 b)
+{
+  __VEC_U_640 result;
+  vui128_t mc, mp, mq;
+
+  result.vx0 = vec_addcq (&mc, a.vx0, b.vx0);
+  result.vx1 = vec_addeq (&mp, a.vx1, b.vx1, mc);
+  result.vx2 = vec_addeq (&mc, a.vx2, b.vx2, mp);
+  result.vx3 = vec_addeq (&result.vx4, a.vx3, b.vx3, mc);
+  return result;
+}
+
+/** \brief Vector Add Extended 512-bit Unsigned Integer & write Carry.
+ *
+ *  Compute the 512 bit sum of two 512 bit values a, b and
+ *  1 bit value carry-in value c.
+ *  Product the carry out of the high order bit of the sum.
+ *  The sum (with-carry) is returned as single 640-bit integer in a
+ *  homogeneous aggregate structure.
+ *
+ *  |processor|Latency|Throughput|
+ *  |--------:|:-----:|:---------|
+ *  |power8   |  16   | 1/cycle  |
+ *  |power9   |  12   | 1/cycle  |
+ *
+ *  @param a vector representation of a unsigned 512-bit integer.
+ *  @param b vector representation of a unsigned 512-bit integer.
+ *  @param c vector representation of a unsigned 1-bit carry.
+ *  @return homogeneous aggregate representation of the unsigned
+ *  640-bit sum of a + b + c.
+ */
+static inline __VEC_U_640
+vec_add512ecu (__VEC_U_512 a, __VEC_U_512 b, vui128_t c)
+{
+  __VEC_U_640 result;
+  vui128_t mp, mq;
+
+  result.vx0 = vec_addeq (&mq, a.vx0, b.vx0, c);
+  result.vx1 = vec_addeq (&mp, a.vx1, b.vx1, mq);
+  result.vx2 = vec_addeq (&mq, a.vx2, b.vx2, mp);
+  result.vx3 = vec_addeq (&result.vx4, a.vx3, b.vx3, mq);
+  return result;
+}
+
+/** \brief Vector Add Extended 512-bit Unsigned Integer Modulo.
+ *
+ *  Compute the 512 bit sum of two 512 bit values a, b and
+ *  1 bit value carry-in value c.
+ *  The sum is returned as single 512-bit integer in a
+ *  homogeneous aggregate structure.
+ *  Any carry-out of the high order bit of the sum is lost.
+ *
+ *  |processor|Latency|Throughput|
+ *  |--------:|:-----:|:---------|
+ *  |power8   |  16   | 1/cycle  |
+ *  |power9   |  12   | 1/cycle  |
+ *
+ *  @param a vector representation of a unsigned 512-bit integer.
+ *  @param b vector representation of a unsigned 512-bit integer.
+ *  @param c vector representation of a unsigned 1-bit carry.
+ *  @return homogeneous aggregate representation of the unsigned
+ *  512-bit sum of a + b + c.
+ 
```

Comment:
It could be helpful to add a comment in the comments just above reminding why the `COMPILE_FENCE`s are here. (there are additional places below where this suggestion applies.)

Reply:
*  \note We use the COMPILER_FENCE to limit instruction scheduling
 *  and code motion to smaller code blocks. This in turn reduces
 *  register pressure to and avoid generating spill.